skyportal/handlers/api/stream.py

```diff
@@ -0,0 +1,168 @@
+        """
+        ---
+        single:
+          description: Retrieve a stream
+          parameters:
+            - in: path
+              name: filter_id
+              required: true
+              schema:
+                type: integer
+          responses:
+            200:
+              content:
+                application/json:
+                  schema: SingleFilter
+            400:
+              content:
+                application/json:
+                  schema: Error
+        multiple:
+          description: Retrieve all streams
+          responses:
+            200:
+              content:
+                application/json:
+                  schema: ArrayOfFilters
+            400:
+              content:
+                application/json:
+                  schema: Error
+        """
+        if stream_id is not None:
+            # fixme: add ACLs! Users should be created with specific Stream access permissions
+            # s = Stream.get_if_owned_by(stream_id, self.current_user)
+            s = DBSession().query(Stream).filter(Stream.id == stream_id).first()
+            if s is None:
+                return self.error("Invalid stream ID.")
+            return self.success(data=s)
+        streams = (
+            DBSession().query(Stream)
+            # fixme: results in error "'Token' object has no attribute 'streams'"
+            # .filter(Stream.id.in_([s.id for s in self.current_user.streams]))
+            .all()
+        )
+        return self.success(data=streams)
+
+    @permissions(["System admin"])
+    def post(self):
+        """
+        ---
+        description: POST a new stream.
+        requestBody:
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  name:
+                    type: string
+                  altdata:
+                    type: object
+                required:
+                  - name
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - $ref: '#/components/schemas/Success'
+                    - type: object
+                      properties:
+                        data:
+                          type: object
+                          properties:
+                            id:
+                              type: integer
+                              description: New stream ID
+        """
+        data = self.get_json()
+        schema = Stream.__schema__()
+        try:
+            stream = schema.load(data)
+        except ValidationError as e:
+            return self.error(
+                "Invalid/missing parameters: " f"{e.normalized_messages()}"
+            )
+        DBSession().add(stream)
+        DBSession().commit()
+
+        return self.success(data={"id": stream.id})
+
+    @permissions(["System admin"])
+    def patch(self, stream_id):
```

Comment:
We have been using PUT for partial updates, although we should be using PATCH according to the HTTP standards. 

@stefanv What do you think is the appropriate path here? Use PUT for now and then open another PR changing all the PUTs to PATCHes?

Reply:
We use PATCH for filter and user preferences updates, so there are precedents :)